src/sign/keyset.rs

```diff
@@ -0,0 +1,1906 @@
+        assert_eq!(zone_sigs(&ks), ["first CSK"]);
+        assert_eq!(ds_keys(&ks), ["first CSK"]);
+
+        let actions = ks.roll_done(RollType::CskRoll).unwrap();
+        assert_eq!(actions, []);
+        ks.delete_key("second KSK").unwrap();
+        ks.delete_key("second ZSK").unwrap();
+
+        ks.add_key_csk("second CSK".to_string(), None, UnixTime::now());
+
+        let actions = ks
+            .start_roll(RollType::CskRoll, &["first CSK"], &["second CSK"])
+            .unwrap();
+        assert_eq!(
+            actions,
+            [Action::UpdateDnskeyRrset, Action::ReportDnskeyPropagated]
+        );
+        assert_eq!(dnskey(&ks), ["first CSK", "second CSK"]);
+        assert_eq!(dnskey_sigs(&ks), ["first CSK", "second CSK"]);
+        assert_eq!(zone_sigs(&ks), ["first CSK"]);
+        assert_eq!(ds_keys(&ks), ["first CSK"]);
+
+        let actions =
+            ks.propagation1_complete(RollType::CskRoll, 3600).unwrap();
+        assert_eq!(actions, []);
+
+        MockClock::advance_system_time(Duration::from_secs(3600));
+
+        println!("CSK roll cache expired1");
+        let actions = ks.cache_expired1(RollType::CskRoll).unwrap();
+        assert_eq!(
+            actions,
+            [
+                Action::CreateCdsRrset,
+                Action::UpdateDsRrset,
+                Action::UpdateRrsig,
+                Action::ReportDsPropagated,
+                Action::ReportRrsigPropagated
+            ]
+        );
+        assert_eq!(dnskey(&ks), ["first CSK", "second CSK"]);
+        assert_eq!(dnskey_sigs(&ks), ["first CSK", "second CSK"]);
+        assert_eq!(zone_sigs(&ks), ["second CSK"]);
+        assert_eq!(ds_keys(&ks), ["second CSK"]);
+
+        let actions =
+            ks.propagation2_complete(RollType::CskRoll, 3600).unwrap();
+        assert_eq!(actions, []);
+
+        MockClock::advance_system_time(Duration::from_secs(3600));
+
+        let actions = ks.cache_expired2(RollType::CskRoll).unwrap();
+        assert_eq!(
+            actions,
+            [Action::RemoveCdsRrset, Action::UpdateDnskeyRrset]
+        );
+        assert_eq!(dnskey(&ks), ["second CSK"]);
+        assert_eq!(dnskey_sigs(&ks), ["second CSK"]);
+        assert_eq!(zone_sigs(&ks), ["second CSK"]);
+        assert_eq!(ds_keys(&ks), ["second CSK"]);
+
+        let actions = ks.roll_done(RollType::CskRoll).unwrap();
+        assert_eq!(actions, []);
+        ks.delete_key("first CSK").unwrap();
+    }
+
+    fn dnskey(ks: &KeySet) -> Vec<String> {
+        let keys = ks.keys();
+        let mut vec = Vec::new();
+        for key in keys {
+            let status = match key.keytype() {
+                KeyType::Ksk(keystate)
+                | KeyType::Zsk(keystate)
+                | KeyType::Csk(keystate, _)
+                | KeyType::Include(keystate) => keystate,
+            };
+            if status.present() {
+                vec.push(key.pubref().to_string());
+            }
+        }
+        vec
```

Comment:
Use `ks.keys.iter().filter().map().collect()`.  It's, simpler, more efficient and more idiomatic.

Reply:
Done